src/serialization.jl

```diff
@@ -58,7 +63,49 @@ function serialize(vals::Vector{T}) where {T <: InfrastructureSystemsType}
 end
 
 function serialize_struct(val::T) where {T}
-    return Dict(string(name) => serialize(getfield(val, name)) for name in fieldnames(T))
+    @debug "serialize_struct" val T
+    data = Dict(string(name) => serialize(getfield(val, name)) for name in fieldnames(T))
+    add_serialization_metadata!(data, T)
+    return data
+end
+
+"""
+Add type information to the dictionary that can be used to deserialize the value.
+"""
+function add_serialization_metadata!(data::Dict, ::Type{T}) where {T}
+    data[METADATA_KEY] = Dict{String, Any}(
+        TYPE_KEY => string(nameof(T)),
+        MODULE_KEY => string(parentmodule(T)),
+    )
+    if !isempty(T.parameters)
+        data[METADATA_KEY][PARAMETERS_KEY] = string.(T.parameters)
+    end
+
+    return
+end
+
+"""
+Return the type information for the serialized struct.
+"""
+get_serialization_metadata(data::Dict) = data[METADATA_KEY]
+
+function get_type_from_serialization_data(data::Dict)
+    return get_type_from_serialization_metadata(get_serialization_metadata(data))
+end
+
+function get_type_from_serialization_metadata(metadata::Dict)
+    mod = Base.root_module(Base.__toplevel__, Symbol(metadata[MODULE_KEY]))
```

Comment:
@kdheepak FYI, in case you're interested.  I implemented the changes we discussed here.  Also note the use of `root_module` here.  It allows us to find modules above the current scope.